distribution/push_v2.go

```diff
@@ -263,6 +265,37 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
 	// then push the blob.
 	bs := pd.repo.Blobs(ctx)
 
+	// attempt to find another repository to mount the layer from
+	for _, blobsum := range possibleBlobsums {
+		sourceRepo, err := reference.ParseNamed(blobsum.SourceRepository)
+		if err != nil {
+			continue
+		}
+		if pd.repoInfo.Hostname() == sourceRepo.Hostname() {
+			logrus.Debugf("attempting to mount layer %s (%s) from %s", diffID, blobsum.Digest, sourceRepo.FullName())
+
+			desc, err := bs.Mount(ctx, sourceRepo.RemoteName(), blobsum.Digest)
+			if err == nil {
+				progress.Updatef(progressOutput, pd.ID(), "Mounted from %s", sourceRepo.RemoteName())
```

Comment:
If that is a case then the appropriate behavior after receiving `StatusAccepted` would be to break so multiple uploads are not created.  Looking at the registry implementation, a `202` is only expected here if the registry does not support cross repo mount, so retrying does not even seem like it could be fruitful.